service/history/shard/context_impl.go

```diff
@@ -2025,7 +2025,7 @@ func (s *ContextImpl) ensureMinContextTimeout(
 		return ctx, func() {}, nil
 	}
 
-	newContext, cancel := context.WithTimeout(context.Background(), minContextTimeout)
+	newContext, cancel := context.WithTimeout(s.lifecycleCtx, minContextTimeout)
```

Comment:
This one is a little weird since sometimes the context returned will be cancellable by lifecycleCtx and sometimes it won't, depending on the deadline. This shouldn't make things _worse_ since it's replacing an uncancellable `Background()` by a cancellable context. But if it feels weird, I can revert this one. The ones in `loadShardMetadata` and `renewRangeLocked` are the ones I want to be cancellable.

Reply:
I'm confused, we don't use passed in ctx as parent?

Reply:
Sometimes yes, sometimes not.. the point is that we can't control ctx and it might have a really small deadline, which makes the operation fail for no good reason, and we reacquire the shard in that case. See description on #2660

Reply:
The timeout in the passed in ctx may to too small (since it comes from the API) and if we use it directly to call persistence, we risk getting a timeout error and need to re-acquire the shard.

Idea of this function is to ensure some minimal timeout when calling persistence api that may lead to shard movement.